### ISSUE TITLE
Is ChangeDetectionStrategy.OnPush used correctly in this app?

### ISSUE THREAD

**Issue by junimohano** [ngrx/example-app#164]
I'm confused when should I use onPush each component..

I searched other examples in terms of ChangeDetectionStrategy, they put onPush in each components not containers, but in this example-app, ChangeDetectionStrategy.OnPush is in containers

Can you check is this correct or not ?

**Comment by piq9117** [comment]
I'm curious to find out as well. 
I thought `ChangeDetectionStrategy.OnPush` depended on `@Input`. Thats why its used in child components.

**Comment by sidewayskumara** [comment]
My guess is that it can be used this way because a) containers are not presentational components and are not relied upon to re-render the view and b) the strategy applies to all child components (https://gist.github.com/btroncone/a6e4347326749f938510#taking-advantage-of-changedetectiononpush) but I would also like confirmation on the convention with this.

**Comment by sharikovvladislav** [comment]
I think developers of example-app used this because:
1) you get advantages of using `ChangeDetectionStrategy.OnPush` (speed)
2) you still get working application because the AsyncPipe

`ChangeDetectionStrategy.OnPush` makes your app faster. If component use this strategy it will compare input by reference (new and old inputs), but not with deep equality. Compare by reference is incredibly faster.

You still can mark component for check (or say Angular that this component has to be re-rendered) by using `ChangeDetectorRef`. You can use `detectChanges` or `markForCheck` methods. If you can see there is no using of `ChangeDetectorRef` service. 

But! AsyncPipe makes this for you. You can read about it there: https://blog.thoughtram.io/angular/2017/02/27/three-things-you-didnt-know-about-the-async-pipe.html . Proof is in the AsyncPipe code: https://github.com/angular/angular/blob/master/packages/common/src/pipes/async_pipe.ts#L142

As you can see in the example app code, they use async in `bc-collection-page`: https://github.com/ngrx/example-app/blob/master/src/app/containers/collection-page.ts#L18. So everything is working (view is updating) correct.

@krjordan can you correct me if I said something wrong?

Everything above is just my assumes :)

**Comment by piq9117** [comment]
`ChangeDetectionStrategy` is not for speed. Its for controlling re-renders. `OnPush` strategy will trigger a re-render if the state changes. If there are no state changes then it will not recompute, and I guess you get that feeling of "speed"

**Comment by sharikovvladislav** [comment]
@piq9117 exactly. it is about re-render. Nothing new on input? No re-render. How Angular get diff? Only by comparing references of new and old inputs. Thats is much more faster then default way. Yes I meant this.

**Comment by sreekumarp** [comment]
So if we have any local public variable changes (other than @Input or @Output variables) we need to call `markForCheck()` method just below we made the change... right ? 

and which is better `markForCheck` or `dectectChanges` ?
